src/clj/puppetlabs/services/master/master_core.clj

```diff
@@ -332,16 +332,30 @@
     (->> (map format-task info-from-jruby)
          (middleware-utils/json-response 200))))
 
+(defn task-file-uri-components
+  "The 'name' portion for a task implementation is a single component like
+  'foo.sh' and can never be nested in subdirectories. In that case we know the
+  'file' must be structured <environment>/<module root>/<module>/tasks/<filename>.
+  Other task files are the path relative to the module root, in the form
+  <module>/<mount>/<path> where <path> may have subdirectories. For those, we
+  just slice that off the end of the file path and take the last component that
+  remains as the module root."
+  [file-name file]
```

Comment:
(nit) I think these parameter names could be a little clearer. When I see `file-name`, I imagine a single component, like you describe in the docstring. I think `subpath`, as is used in `describe-task-file` below, would be clearer. I would also probably call `file` something like `full-path`, but I think I'd rather you maintain the consistency with `describe-task-file` than just change this one.

Reply:
I updated this to go with `file-id` in both places, as that reflects what that field actually means. It's supposed to specify _which_ file should be used, separate from _where_ the file is located. Bolt calls it `name` but I agree that `file-name` is ambiguous, so `id` seems like a decent compromise.

I _don't_ like `subpath` because it's not actually a "sub" of the absolute path. It has its own semantics.